File: src/services/ReceiptProcessor.js
const { spawn } = require('child_process');
const fs = require('fs').promises;
const path = require('path');

class ReceiptProcessor {
  constructor() {
    this.pythonScript = path.join(__dirname, 'receipt_processor.py');
  }

  /**
   * Process a receipt image using EasyOCR
   * @param {string} imagePath - Path to the receipt image
   * @returns {Promise<Object>} Processed receipt data
   */
  async processReceiptImage(imagePath) {
    try {
      console.log('🔍 Processing receipt image with EasyOCR...');

      // Check if image file exists
      await fs.access(imagePath);

      // Run Python OCR processor
      const result = await this.runPythonProcessor(imagePath);

      if (result.error) {
        throw new Error(result.error);
      }

      console.log('✅ Receipt processed successfully with EasyOCR');
      console.log(`📄 Extracted ${result.processed_text?.length || 0} characters using ${result.extraction_method}`);

      return result;

    } catch (error) {
      console.error('❌ Receipt processing failed:', error.message);
      throw error;
    }
  }

  /**
   * Run the Python receipt processor
   * @param {string} imagePath - Path to image
   * @returns {Promise<Object>} Processing result
   */
  runPythonProcessor(imagePath) {
    return new Promise((resolve, reject) => {
      const python = spawn('python', [this.pythonScript, imagePath]);

      let stdout = '';
      let stderr = '';

      python.stdout.on('data', (data) => {
        stdout += data.toString();
      });

      python.stderr.on('data', (data) => {
        stderr += data.toString();
      });

      python.on('close', (code) => {
        if (stderr) {
          console.log('Python processor logs:', stderr);
        }

        if (code !== 0) {
          reject(new Error(`Python processor failed with code ${code}`));
          return;
        }

        try {
          const result = JSON.parse(stdout);
          resolve(result);
        } catch (parseError) {
          reject(new Error(`Failed to parse processor output: ${parseError.message}`));
        }
      });

      python.on('error', (error) => {
        reject(new Error(`Failed to start Python processor: ${error.message}`));
      });
    });
  }

  /**
   * Create a structured prompt for AI processing
   * @param {Object} processedData - Data from EasyOCR
   * @returns {string} Formatted prompt for AI
   */
  createAIPrompt(processedData) {
    const prompt = `
Please analyze this receipt/payment text extracted via OCR and extract transaction information in JSON format.

${processedData.processed_text}

Extract the following information:
- description: ONLY if clearly identifiable from the text (e.g., specific item names, service descriptions). If unclear or generic, leave as null or empty string.
- amount: Total amount (number only, no currency symbols)
- merchant: Store/business name or payment app (PayLah, GrabPay, etc.)
- category: Expense category (Food & Dining, Transportation, Shopping, Transfer, etc.)
- date: Transaction date if visible (YYYY-MM-DD format)
- payment_method: Payment method used (Cash, Card, PayLah, GrabPay, etc.)

IMPORTANT: Do NOT generate generic descriptions like "Receipt transaction" or "Purchase". Only include a description if you can identify specific items or services from the receipt text.

Return only valid JSON in this format:
{
  "description": null,
  "amount": 0.00,
  "merchant": "...",
  "category": "...",
  "date": "...",
  "payment_method": "...",
  "confidence": 0.9
}`;

    return prompt;
  }

  /**
   * Test the EasyOCR processor with a sample image
   * @param {string} imagePath - Path to test image
   */
  async testProcessor(imagePath) {
    try {
      console.log('🧪 Testing EasyOCR receipt processor...');
      const result = await this.processReceiptImage(imagePath);

      console.log('📊 OCR Results:');
      console.log('- Method:', result.extraction_method);
      console.log('- Success:', result.success);
      console.log('- Text length:', result.processed_text?.length || 0);
      console.log('- Confidence details:', result.confidence_details?.length || 0, 'text regions');

      if (result.processed_text) {
        console.log('\n📄 Extracted Text Preview:');
        console.log(result.processed_text.substring(0, 300) + '...');
      }

      return result;

    } catch (error) {
      console.error('❌ EasyOCR test failed:', error.message);
      throw error;
    }
  }
}

module.exports = ReceiptProcessor;

File: src/services/receipt_processor.py
#!/usr/bin/env python3
"""
Receipt Processing Service using EasyOCR
Converts receipt images to structured text for better AI understanding
"""

import sys
import json
import os
from pathlib import Path

try:
    import easyocr
except ImportError as e:
    print(json.dumps({"error": f"Missing EasyOCR: {e}"}))
    sys.exit(1)

def process_receipt_image(image_path):
    """
    Process a receipt image using EasyOCR
    Returns structured text data
    """
    try:
        print(f"Processing image with EasyOCR: {image_path}", file=sys.stderr)
        
        # Check if file exists and is readable
        if not os.path.exists(image_path):
            return {"error": f"Image file not found: {image_path}"}
        
        # Initialize EasyOCR reader
        reader = easyocr.Reader(['en'], gpu=False)  # Force CPU to avoid GPU issues
        
        # Extract text from image
        ocr_results = reader.readtext(image_path)
        
        if not ocr_results:
            return {"error": "No text detected in image"}
        
        # Filter results by confidence and extract text
        high_confidence_text = []
        all_text = []
        
        for result in ocr_results:
            bbox, text, confidence = result
            all_text.append(f"{text} (conf: {confidence:.2f})")
            
            # Only include high confidence text in main extraction
            if confidence > 0.5:
                high_confidence_text.append(text)
        
        if not high_confidence_text:
            return {"error": "No high-confidence text detected in image"}
        
        # Join text with newlines to preserve structure
        extracted_text = "\n".join(high_confidence_text)
        
        print(f"EasyOCR extracted {len(extracted_text)} characters from {len(high_confidence_text)} text regions", file=sys.stderr)
        
        return {
            "raw_text": extracted_text,
            "processed_text": clean_receipt_text(extracted_text),
            "extraction_method": "easyocr",
            "confidence_details": all_text,
            "success": True
        }
        
    except Exception as e:
        print(f"EasyOCR processing failed: {e}", file=sys.stderr)
        return {"error": f"OCR processing failed: {str(e)}"}

def clean_receipt_text(text):
    """
    Clean and structure the extracted text for better AI processing
    """
    if not text:
        return "No text extracted"
    
    lines = text.strip().split('\n')
    cleaned_lines = []
    
    for line in lines:
        line = line.strip()
        # Skip empty lines, single characters, and common OCR noise
        if line and len(line) > 1 and not line.isspace():
            # Remove excessive whitespace
            line = ' '.join(line.split())
            # Skip lines that are just symbols or numbers without context
            if not line.replace('-', '').replace('_', '').replace('*', '').strip():
                continue
            cleaned_lines.append(line)
    
    if not cleaned_lines:
        return "No meaningful text extracted from receipt"
    
    # Structure the text with clear sections
    structured_text = "RECEIPT CONTENT:\n"
    structured_text += "=" * 40 + "\n"
    
    # Group lines that might be related
    for i, line in enumerate(cleaned_lines):
        structured_text += f"{line}\n"
        
        # Add spacing between different sections (heuristic)
        if i < len(cleaned_lines) - 1:
            current_line = line.lower()
            next_line = cleaned_lines[i + 1].lower()
            
            # Add spacing after store names, dates, totals
            if any(keyword in current_line for keyword in ['total', 'subtotal', 'tax', 'date', 'time']):
                structured_text += "\n"
    
    structured_text += "=" * 40 + "\n"
    structured_text += "END OF RECEIPT"
    
    return structured_text

def main():
    if len(sys.argv) != 2:
        print(json.dumps({"error": "Usage: python receipt_processor.py <image_path>"}))
        sys.exit(1)
    
    image_path = sys.argv[1]
    
    if not os.path.exists(image_path):
        print(json.dumps({"error": f"Image file not found: {image_path}"}))
        sys.exit(1)
    
    result = process_receipt_image(image_path)
    print(json.dumps(result, indent=2))

if __name__ == "__main__":
    main()